benchmark/C/Savina/src/micro/Throughput.lf

```diff
@@ -0,0 +1,183 @@
+ * batch processing messages in the worker actors.
+ * This is not possible with reactors because scheduling and
+ * advancing logical time occurs after each round of
+ * message delivery.
+ * 
+ * @author Hannes Klein (c++ version)
+ * @author Matthew Chorlian (adapted C++ version to C)
+ */
+
+
+/* [[[cog
+# This file is a code generator using the python module cog:
+# See https://nedbatchelder.com/code/cog/
+#
+# All instructions for code  generation are in-lined in comments
+# like this one. With that you can use this file as a normal source file
+# but also to generate code.
+# 
+# To change the generated code in-line within this file run:
+# $ python -m cog -r this-file.lf
+# To generate a new file from this file stripping the generator code in the process run:
+# $ python -m cog -d -o output-file.lf this-file.lf
+#
+# Use the command line option -D to specify generator parameters, for example:
+# $ python -m cog -r -D parameter=100 this-file.lf
+#
+# Generator parameters used in this file:
+# -D numWorkers=60
+# 
+]]] */
+// [[[end]]]
+
+/* [[[cog
+  # force existence, type and default values of generator parameters
+  if 'numWorkers' in globals():
+    numWorkers = int(numWorkers)
+  else:
+    globals()['numWorkers'] = 60
+  
+  # output the current value of the generator parameters used in the last generation run
+  cog.outl(f'// Generated file with the following parameters:')
+  cog.outl(f'// numWorkers = {numWorkers}')
+]]] */
+// Generated file with the following parameters:
+// numWorkers = 60
+// [[[end]]]
+
+target C {
+    flags: "-lm",
+    fast: true
+};
+
+
+reactor ThroughputReactor(totalMessages:int(10000)) {
+    
+    preamble {=
+        #include <math.h>
+
+        void performComputation(double theta) {
+            double sint = sin(theta);
+            double res = sint * sint;
+            // defeat dead code elimination
+            if(res <= 0) {
+                printf("Benchmark exited with unrealistic res value.");
+            }
+        }
+    =}
+    
+    input inReset:bool;
+    input inMessage:bool;
+    output outFinished:bool;
+    
+    state messagesProcessed:int(0);
+    
+    reaction(inReset) {=
+        self->messagesProcessed = 0;
+    =}
+    
+    reaction(inMessage) -> outFinished {=
+        self->messagesProcessed++;
+        performComputation(37.2);
+        
+        if(self->messagesProcessed == self->totalMessages) {
+            SET(outFinished, true);
+        }
+    =}
+    
+}
+
+reactor ThroughputProducer(totalMessages:int(10000), numConsumer:int(60)) {
+    
+    state messagesSent:int(0);
+    state nextConsumer:int(0);
+    state numConsumerFinished:int(0);
+    
+    input inStart:bool;
+    output outFinished:bool;
+  
+     /* [[[cog
+      cog.outl(f'output[{numWorkers}] outMessage:bool;')
+      cog.outl(f'input[{numWorkers}] inFinished:bool;')
+      cog.outl(f'output[{numWorkers}] outReset:bool;')
+    ]]] */
+    output[60] outMessage:bool;
```

Comment:
This can be simplified a bit by using the reactor parameters. For instance, you can write `output[numConsumer] outMessage:bool`. Note that this is a relatively new feature, that did not exist when the C++ benchmarks where written. We are currently updating all the C++ benchmarks in #406.